### ISSUE TITLE
Unable to determine IP address of 'eth0' when specified for VirtualHost

### ISSUE THREAD

**Issue by mdell-seradex** [proftpd/proftpd#1282]
### What I Did
In an attempt to de-couple the ftp server from the IP address as it is likely to be moved soon, I tried to setup the VirtualHost to be based on the interface name instead of a fixed IP address.  
Instead of it working as desired, I am unable to the ftp server with this setup, and when I run `sudo systemctl status proftpd`  I get that warning. `proftpd[139961] FTP: warning: unable to determine IP address of 'eth0'`
Unfortunately I can not find much documentation on this functionality.   I do see in the http://www.proftpd.org/docs/NEWS-1.3.6e that this functionality was added to address [Bug #3820](http://bugs.proftpd.org/show_bug.cgi?id=3820).  There was other mentions of this that I had found as well, but I am unsure where at this time.
When I replace `eth0` with the actual IP address, everything works as expected.
Please note that this is not urgent since I was simply trying to make things easier for myself.  I do not expect the IP address to change often or that the ftp server will be relocated frequently.

### ProFTPD Version and Configuration
Ubuntu 20.04 LTS HyperV image - ProFTPD Version 1.3.6c installed using `sudo apt-get install proftpd`.
A custom compilation was not used.
Additional info can be provided if required.  Please do not hesitate to ask.

Output from `proftpd -V`:
```
Compile-time Settings:
  Version: 1.3.6c (maint)
  Platform: LINUX [Linux 5.8.0-53-generic x86_64]
  Built: Thu Feb 27 2020 19:34:56 UTC
  Built With:
    configure  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--with-includes=/usr/include/postgresql:/usr/include/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g2 -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-oI4YZ3/proftpd-dfsg-1.3.6c=. -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -rdynamic  -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu
  LIBS: -lacl  -lpcreposix -lpcre -lssl -lcrypto -lcap  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -ldl -lhiredis -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + Redis support
    - Sodium support
    + OpenSSL support
    + PCRE support
    + POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

proftpd.conf:
```
# Includes DSO modules - default list included in package.
Include /etc/proftpd/modules.conf

# Set off to disable IPv6 support which is annoying on IPv4 only boxes.
IdentLookups                   off

ServerType                     standalone
DeferWelcome                   off

MultilineRFC2228               on
DefaultServer                  on
ShowSymlinks                   on

TimeoutNoTransfer              600
TimeoutStalled                 600
TimeoutIdle                    1200

DisplayLogin                   welcome.msg
DisplayChdir                   .message true
ListOptions                    "-l"

DenyFilter                     \*.*/

<IfModule mod_dso.c>
  LoadModule mod_case.c
</IfModule>

# Use this to jail all users in their homes 
DefaultRoot                    ~

# Port 21 is the standard FTP port.  Port 0 disables the generic setup
Port                           0

# Ensures that a TCP listening socket is created and bound to a specific IP address for the main "server config" server and all configured virtual servers instead of a wildcard binding to a specific port.
SocketBindTight                on

# To prevent DoS attacks, set the maximum number of child processes to 30.  If you need to allow more than 30 concurrent connections at once, simply increase this value.
# Note that this ONLY works in standalone mode, in inetd mode you should use an inetd server that allows you to limit maximum number of processes per service (such as xinetd)
MaxInstances                   30

# Set the user and group that the server normally runs at.
User                           proftpd
Group                          nogroup

# Umask 022 is a good standard umask to prevent new files and dirs (second parm) from being group and world writable.
Umask                          022  022
# Normally, we want files to be overwriteable.
AllowOverwrite                 on

SysLogLevel debug
SystemLog   /var/log/proftpd/proftpd.log

# In order to keep log file dates consistent after chroot, use timezone info from /etc/localtime.
# If this is not set, and proftpd is configured to chroot (e.g. DefaultRoot or <Anonymous>), it will use the non-daylight savings timezone regardless of whether DST is in effect.
SetEnv TZ :/etc/localtime

<IfModule mod_quotatab.c>
QuotaEngine off
</IfModule>

<IfModule mod_ratio.c>
Ratios off
</IfModule>

# Delay engine reduces impact of the so-called Timing Attack described in http://www.securityfocus.com/bid/11430/discuss. It is on by default. 
<IfModule mod_delay.c>
DelayEngine on
</IfModule>

<IfModule mod_ctrls.c>
ControlsEngine        off
ControlsMaxClients    2
ControlsLog           /var/log/proftpd/controls.log
ControlsInterval      5
ControlsSocket        /var/run/proftpd/proftpd.sock
</IfModule>

<IfModule mod_ctrls_admin.c>
AdminControlsEngine off
</IfModule>

# Useful to keep VirtualHost/VirtualRoot directives separated
Include /etc/proftpd/virtuals.conf

# Include other custom configuration files !! Please note, that this statement will read /all/ file from this subdir, i.e. backup files created by your editor, too !!!
# Eventually create file patterns like this: "/etc/proftpd/conf.d/*.conf".  This directory is currently empty.
Include /etc/proftpd/conf.d/
```

virtuals.conf:
```
<VirtualHost eth0>
  ServerName              "Custom name"
  Port                    21
  TransferLog             /var/log/proftpd/xfer/testftpserver.log
  MaxLoginAttempts        1
  RequireValidShell       no
  DefaultRoot             /ftp
  AllowOverwrite          no
  Protocols               ftp
  # Use only AuthUserFiles when authenticating, and not the system's /etc/passwd
  AuthOrder               mod_auth_file.c
  AuthGroupFile           /etc/proftpd/ftpd.group
  AuthUserFile            /etc/proftpd/ftpd.passwd
  DebugLevel              10

  <Directory ~>
    HideNoAccess on
    <Limit WRITE CDUP MDTM MLSD MLST NLST PWD RNFR XCUP XCWD XPWD MFMT SITE_CHGRP SITE_CHMOD SITE_MKDIR SITE_RMDIR SITE_SYMLINK SITE_UTIME>
      DenyAll
    </Limit>
  </Directory>

  <IfModule mod_rewrite.c>
    RewriteEngine on

    # Define a map that uses the internal "replaceall" function
    RewriteMap replace int:replaceall

    # We only want to use this rule on RETR and SIZE commands
    RewriteCondition %m RETR|SIZE|LIST

    # Apply the map to the command parameters
    RewriteRule (.*) "${replace:!$1!\\!/}"
  </IfModule>

  <IfModule mod_case.c>
    CaseEngine on

    # Enable case-insensitivity only for downloads
    CaseIgnore RETR,SIZE,LIST
    CaseLog /var/log/proftpd/proftpd-case.log
  </IfModule>
</VirtualHost>
```

Output from `ifconfig eth0`:
```
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.30.118.36  netmask 255.255.240.0  broadcast 172.30.127.255
        inet6 fe80::29ff:47bd:46db:69fd  prefixlen 64  scopeid 0x20<link>
        ether 00:15:5d:13:11:0a  txqueuelen 1000  (Ethernet)
        RX packets 2363336  bytes 1150450625 (1.1 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1348415  bytes 1837555043 (1.8 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Comment by Castaglia** [comment]
I have reproduced this issue locally, using an Ubuntu 20.04 Docker image.  To obtain more details, I enabled trace logging like so:
```
TraceLog /var/log/proftpd/trace.log
Trace dns:30
```
In the generated trace logging, I see:
```
2021-08-07 16:07:37,076 [7867] <dns:10>: resolving name 'eth0' to IP address
2021-08-07 16:07:37,076 [7867] <dns:7>: attempting to resolve 'eth0' to IPv4 address via DNS
2021-08-07 16:07:37,094 [7867] <dns:1>: IPv4 getaddrinfo 'eth0' error: Name or service not known
2021-08-07 16:07:37,094 [7867] <dns:7>: resolved 'eth0' to interface with IPv4 address 172.18.0.2
```
which suggests that the implementation works as desired.

However, later, I see:
```
2021-08-07 16:07:37,097 [7867] <dns:10>: resolving name 'eth0' to IP address
2021-08-07 16:07:37,097 [7867] <dns:2>: no IP address found in netaddr IP cache for name 'eth0'
2021-08-07 16:07:37,097 [7867] <dns:7>: attempting to resolve 'eth0' to IPv4 address via DNS
2021-08-07 16:07:37,098 [7867] <dns:1>: IPv4 getaddrinfo 'eth0' error: Name or service not known
2021-08-07 16:07:37,098 [7867] <dns:8>: failed to resolve 'eth0' to an IP address
```
So something is not quite right.

**Comment by Castaglia** [comment]
Fixed now in master, and backported to the 1.3.7 branch.  Thanks!

**Comment by mdell-seradex** [comment]
Thank you for finding and addressing it @Castaglia. 👍
FYI, I had this issue on v1.3.6c and see this fix was not backported there, however, I don't know if you are supporting that version any more.
Also, while the news says this is now supported, the documentation generally does not. I'll try to put together a pull request to address the documentation.

**Comment by Castaglia** [comment]
No, there won't be any backport of the fix to the 1.3.6 branch; at any given time, we support the current development branch (master, the RC releases), and most recent maintenance branch (1.3.7, in this case); see the [Versioning doc](http://www.proftpd.org/docs/howto/Versioning.html).

**Comment by mdell-seradex** [comment]
I understand.  Thank you for that info.
